### PR TITLE
refactor(ZNTA-1017): translation updates event to handle batch

### PR DESCRIPTION
**Pull request by aeng**
https://zanata.atlassian.net/browse/ZNTA-1017

### CODE REVIEW

**Comment by aeng** [comment]
@zanata-jenkins retest this please

**Comment by seanf** [comment]
@aeng There was a NullPointerException in TranslationUpdatedManagerTest.onTranslationUpdateTest

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java [hunk 1]
```diff
@@ -184,24 +185,34 @@ public Object getEntity(EntityType entityType, long entityId) {
...
-                    logActivityAlreadyLocked(actorId,
-                            document.getProjectIteration(), target, activityType,
-                            target.getTextFlow().getWordCount().intValue());
-                });
+                HDocument document =
+                    documentDAO.getById(event.getKey().getDocumentId());
+
+                for (TextFlowTargetStateEvent.TextFlowTargetState state : event
+                    .getStates()) {
+                    transactionUtil.run(() -> {
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: What would happen if we put this transaction around the for loop, instead of inside it?

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java [hunk 2]
```diff
@@ -184,24 +185,34 @@ public Object getEntity(EntityType entityType, long entityId) {
...
+                        HTextFlowTarget target =
+                            textFlowTargetDAO.findById(
+                                state.getTextFlowTargetId(), false);
+                        ActivityType activityType =
+                            state.getNewState().isReviewed()
+                                ? ActivityType.REVIEWED_TRANSLATION
+                                : ActivityType.UPDATE_TRANSLATION;
+
+                        logActivityAlreadyLocked(actorId,
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: I think we should be able to roll up all the activity for a single document (ie a single TextFlowTargetStateEvent), thus generating a single Activity outright instead of generating an Activity, then updating it repeatedly. Should reduce DB activity quite a bit.

File: zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java [hunk 3]
```diff
@@ -237,28 +243,36 @@ private Integer mergeTranslations(
...
-                                        targetVersionId,
-                                        targetTf.getDocument().getId(),
-                                        targetTf.getId(),
-                                        updatedTarget.getLocale().getLocaleId(),
-                                        updatedTarget.getId(),
-                                        updatedTarget.getState(),
-                                        entry.getValue());
-
-                        textFlowTargetStateEventEvent.fire(event);
+                        DocumentLocaleKey key = new DocumentLocaleKey(null,
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java]: Why the null? We should know which user triggered the merge.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java [hunk 4]
```diff
@@ -256,10 +271,21 @@
             result.translatedTextFlowTarget = hTextFlowTarget;
             results.add(result);
         }
-
+        fireTargetUpdateEvent(eventMap);
         return results;
     }
 
+    private void fireTargetUpdateEvent(
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java]: `fireTargetUpdateEvents` would be better.

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java (hunk 2 above)
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: The activity type is `ActivityType.UPDATE_TRANSLATION` which expect single translation. If we want to refactor this, then it should be another pull request since it will affect the dashboard display.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java [hunk 5]
```diff
@@ -186,6 +190,9 @@
         validateReviewPermissionIfApplicable(translationRequests,
                 projectIteration, hLocale);
 
+        Map<DocumentLocaleKey, List<TextFlowTargetState>> eventMap =
+            Maps.newHashMap();
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java]: I think the code relating to eventMap would be more readable if you were to use a MultiMap.

File: zanata-war/src/main/java/org/zanata/events/DocumentLocaleKey.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * @author Alex Eng<a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
+ */
+@AllArgsConstructor
+@EqualsAndHashCode
+@Getter
+public class DocumentLocaleKey implements Serializable{
+
+    // this may be null in the case of document uploads
+    private final @Nullable Long actorId;
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/events/DocumentLocaleKey.java]: This field doesn't fit with the name of the class, but rather than rename the class, would it be possible to factor it out of here? A single TextFlowTargetStateEvent should only be for a single actor.

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java (hunk 2 above)
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: But don't we already roll up all Activity records for a single document within a period of time (15 minutes or similar)? The end result is a single Activity: x translations were updated, in this period of time, by this user, ending with this translation. With the rolled up events, we should be able to generate that lot in one go.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java (hunk 5 above)
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java]: In fact, we shouldn't even need a map here. Within this method, every value in DocumentLocaleKey is a constant, so we should only need a single TextFlowTargetStateEvent.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java [hunk 7]
```diff
@@ -191,18 +191,22 @@ public Boolean textFlowTargetHasWarningOrError(Long targetId,
...
-        // update document status information
-        updateDocStatusCache(key, event.getTextFlowTargetId());
+        for(TextFlowTargetStateEvent.TextFlowTargetState state: event.getStates()) {
+            Long tftId = state.getTextFlowTargetId();
 
-        // invalidate target validation
-        targetValidationCache.remove(event.getTextFlowTargetId());
+            // update document status information
+            updateDocStatusCache(new DocumentLocaleKey(documentId,
+                localeId), tftId);
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java]: We should only need to update docStatusCache once per TextFlowTargetStateEvent, much the same as we should be able to generate a rolled-up Activity directly from TextFlowTargetStateEvent.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationUpdatedManager.java [hunk 8]
```diff
@@ -55,23 +61,28 @@ public void textFlowStateUpdated(
...
-                            event.getProjectIterationId(),
-                            event.getDocumentId(), event.getLocaleId(),
-                            wordCount,
-                            event.getPreviousState(), event.getNewState())
-            );
+            for(TextFlowTargetState state: event.getStates()) {
+                int wordCount =
+                    textFlowDAO.getWordCount(state.getTextFlowId());
+                documentStatisticUpdatedEvent
+                    .fire(new DocumentStatisticUpdatedEvent(
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationUpdatedManager.java]: We should be able to reduce the number of DocumentStatisticUpdatedEvents now that we have aggregation. We should put a todo here, perhaps something like:

    // TODO turn DocumentStatisticUpdatedEvent into an aggregate event like TextFlowTargetStateEvent
    // TODO generate one DocumentStatisticUpdatedEvent per TextFlowTargetStateEvent

File: zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java [hunk 9]
```diff
@@ -90,19 +93,21 @@ public void create() {
...
+    public void textFlowStateUpdated(
+        @Observes(during = TransactionPhase.AFTER_SUCCESS)
+            TextFlowTargetStateEvent event) {
         VersionLocaleKey key =
-                new VersionLocaleKey(event.getProjectIterationId(),
-                        event.getLocaleId());
+            new VersionLocaleKey(event.getKey().getProjectIterationId(),
+                event.getKey().getLocaleId());
         WordStatistic stats = versionStatisticCache.get(key);
-        if (stats != null) {
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java]: Is it still possible for stats to be null?

File: zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java [hunk 10]
```diff
@@ -90,19 +93,21 @@ public void create() {
...
-                    textFlow.getWordCount().intValue());
+        TextFlowDAO textFlowDAO = serviceLocator.getInstance(TextFlowDAO.class);
+
+        for (TextFlowTargetStateEvent.TextFlowTargetState state : event
+            .getStates()) {
+            int wordCount =
+                textFlowDAO.getWordCount(state.getTextFlowId());
+            stats.decrement(state.getPreviousState(), wordCount);
+            stats.increment(state.getNewState(), wordCount);
             versionStatisticCache.put(key, stats);
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java]: We don't need to do this cache.put inside the for loop.

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java [hunk 11]
```diff
@@ -195,9 +195,9 @@ public void logTextFlowStateUpdate(@Observes(during = TransactionPhase.AFTER_SUC
                 HDocument document =
                     documentDAO.getById(event.getKey().getDocumentId());
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: To be on the safe side, I would move this back inside the transaction the way it used to be.

File: zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java (hunk 9 above)
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/VersionStateCacheImpl.java]: possible if version page is not loaded before.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java (hunk 7 above)
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java]: We need to pass in tftId into updateDocStatusCache. The cache needs to update `lastModifiedBy` information for that document by using target.getLastmodified.

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java (hunk 2 above)
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: We do rollup (Every hour) if its the same EntityId which can be - HDoc, HProjectIter and HTFT.
For this case, the tft is all different.

Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: @aeng The TFT is different each time, but ultimately we only store one Activity with (a) the total number of words translated and (b) a pointer to the last TFT. Don't we? That was the intention. 

If we're really generating one Activity for every TextFlowTarget (rolling up only when the TFT is modified within the hour), it's no wonder Activity is such a performance problem for us. How do all these Activities appear in the dashboard, eg when a translator translates 20 TFTs in the same document in an hour? Is that shown as 20 lines in the dashboard?

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java (hunk 7 above)
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java]: So why not pass in the last TFT details from the ImmutableList, and use that to update docStatusCache? All the TFTs should have the same timestamp and account if they're part of a batch, but using the last one's metadata should give the desired effect.

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java (hunk 2 above)
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: > If we're really generating one Activity for every TextFlowTarget (rolling up only when the TFT is modified within the hour),

No. The entry in activity table will be created/updated based on version id, type of activity (eg. update translation) and rolled up within the hour.

>  How do all these Activities appear in the dashboard, eg when a translator translates 20 TFTs in the same document in an hour? Is that shown as 20 lines in the dashboard?

If you constantly update translations in the same version, then its only 1 entry in the activity table.

Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: But we do need the targetId to record against that entry as the last translated string.

Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: I think we can use the last TFT details in what you suggested in `TranslationStateCacheImpl`. What do you think?

Comment (aeng) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: But then `translation` and `review` is 2 different activity type. We can try to aggregate these 2 types of activity, total up the word counts and fire 2 event.

Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: Sounds good.

On 15 April 2016 at 15:27, Alex Eng <notifications@github.com> wrote:

> In
> zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java
> <https://github.com/zanata/zanata-server/pull/1138#discussion_r59828360>:
>
> > -                });
> > +                HDocument document =
> > +                    documentDAO.getById(event.getKey().getDocumentId());
> > +
> > +                for (TextFlowTargetStateEvent.TextFlowTargetState state : event
> > +                    .getStates()) {
> > +                    transactionUtil.run(() -> {
> > +                        HTextFlowTarget target =
> > +                            textFlowTargetDAO.findById(
> > +                                state.getTextFlowTargetId(), false);
> > +                        ActivityType activityType =
> > +                            state.getNewState().isReviewed()
> > +                                ? ActivityType.REVIEWED_TRANSLATION
> > +                                : ActivityType.UPDATE_TRANSLATION;
> > +
> > +                        logActivityAlreadyLocked(actorId,
>
> But then translation and review is 2 different activity type. We can try
> to aggregate these 2 types of activity, total up the word counts and fire 2
> event.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/1138/files/462a372aa89c14a002268b729deeb53d52a19f33#r59828360>
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationUpdatedManager.java [hunk 12]
```diff
@@ -55,23 +61,30 @@ public void textFlowStateUpdated(
...
-                            event.getProjectIterationId(),
-                            event.getDocumentId(), event.getLocaleId(),
-                            wordCount,
-                            event.getPreviousState(), event.getNewState())
-            );
+            for(TextFlowTargetState state: event.getStates()) {
+                int wordCount =
+                    textFlowDAO.getWordCount(state.getTextFlowId());
+                // TODO: turn DocumentStatisticUpdatedEvent into an aggregate event like TextFlowTargetStateEvent
+                // TODO: generate one DocumentStatisticUpdatedEvent per TextFlowTargetStateEvent
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationUpdatedManager.java]: Sorry, can you make that TODO PERF? It looks like DocumentServiceImpl.documentStatisticUpdated is the only observer for DocumentStatisticUpdatedEvent, and it would really benefit from batching, especially for projects which have webhooks enabled.

// TODO PERF: generate one DocumentStatisticUpdatedEvent per TextFlowTargetStateEvent. See DocumentServiceImpl.documentStatisticUpdated

File: zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java [hunk 13]
```diff
@@ -184,24 +187,56 @@ public Object getEntity(EntityType entityType, long entityId) {
...
-                            document.getProjectIteration(), target, activityType,
-                            target.getTextFlow().getWordCount().intValue());
+                    HDocument document =
+                        documentDAO.getById(event.getKey().getDocumentId());
+
+                    HTextFlowTarget lastReviewedTarget = null;
+                    HTextFlowTarget lastTranslatedTarget = null;
+
+                    int totalReviewedWord = 0;
+                    int totalTranslatedWord = 0;
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java]: Plurals (words) would be better. Thanks.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java [hunk 14]
```diff
@@ -859,31 +878,36 @@ public String apply(TextFlowTarget input) {
...
-
             if (handleOp.isPresent()) {
                 handleOp.get().increaseProgress(1);
             }
         }
+        Long actorId =
+            assignCreditToUploader ? authenticatedAccount.getPerson().getId() :
+                null;
+        for (Map.Entry<DocumentLocaleKey, Collection<TextFlowTargetState>> entry : eventMap
+            .asMap().entrySet()) {
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java]: This method (saveBatch) works on a single HLocale and HDocument, so there should only be a single TextFlowTargetStateEvent. We shouldn't need the Map/MultiMap here either.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java [hunk 15]
```diff
@@ -191,18 +191,23 @@ public Boolean textFlowTargetHasWarningOrError(Long targetId,
...
-        // update document status information
-        updateDocStatusCache(key, event.getTextFlowTargetId());
+        for(TextFlowTargetStateEvent.TextFlowTargetState state: event.getStates()) {
+            // invalidate target validation
+            targetValidationCache.remove(state.getTextFlowTargetId());
+        }
 
-        // invalidate target validation
-        targetValidationCache.remove(event.getTextFlowTargetId());
+        Long tftId = event.getStates().get(event.getStates().size() - 1)
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java]: Iterables.getLast from Guava would read more nicely (esp. with static import).

File: docs/release-notes.md [hunk 16]
```diff
@@ -32,6 +32,7 @@
 * [ZNTA-905](https://zanata.atlassian.net/browse/ZNTA-905) - Remove 0% matching translation memory entry
 * [ZNTA-928](https://zanata.atlassian.net/browse/ZNTA-928) - Readonly project doesn't have "lock" icon in UI
 * [ZNTA-793](https://zanata.atlassian.net/browse/ZNTA-793) - Remove unused method in TranslationMemoryAction
+* [ZNTA-1017](https://zanata.atlassian.net/browse/ZNTA-1017) - Refactor translation update event
```
Comment (seanf) [inline docs/release-notes.md]: I don't think refactoring tasks like ZNTA-1017 or ZNTA-793 should appear in the release notes.

File: zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java [hunk 17]
```diff
@@ -186,6 +192,12 @@
         validateReviewPermissionIfApplicable(translationRequests,
                 projectIteration, hLocale);
 
+        DocumentLocaleKey documentLocaleKey =
+            new DocumentLocaleKey(projectIteration.getId(),
+                sampleHTextFlow.getDocument().getId(), hLocale.getLocaleId());
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java]: This `documentLocaleKey` definition could be moved down to just before `new TextFlowTargetStateEvent`

File: zanata-war/src/main/java/org/zanata/events/DocumentLocaleKey.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.io.Serializable;
+
+/**
+ * @author Alex Eng<a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
+ */
+@AllArgsConstructor
+@EqualsAndHashCode
+@Getter
+public class DocumentLocaleKey implements Serializable {
```
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/events/DocumentLocaleKey.java]: - [ ] We already have two other classes named `DocumentLocaleKey` - the only remaining difference between this and the other two is that this one includes the somewhat redundant field `projectIterationId`. Is there any way we can rationalise them all?

**Comment by seanf** [comment]
:+1:

**Comment by djansen-redhat** [comment]
After running copyTrans and seeing a couple of items updated in the logs, the version looks like this:
screenshot from 2016-04-20 14-19-48

**Comment by djansen-redhat** [comment]
:white_check_mark: Looks fine to me
